File: api-legacy/src/main/java/com/bitstachio/bazaarapi/user/service/UserServiceImpl.java
package com.bitstachio.bazaarapi.user.service;

import com.bitstachio.bazaarapi.user.domain.User;
import com.bitstachio.bazaarapi.user.dto.UserCreateRequest;
import com.bitstachio.bazaarapi.user.dto.UserResponse;
import com.bitstachio.bazaarapi.user.dto.UserUpdateRequest;
import com.bitstachio.bazaarapi.user.exception.UserNotFoundException;
import com.bitstachio.bazaarapi.user.repository.UserRepository;

import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class UserServiceImpl implements UserService {

    private final UserRepository userRepository;

    @Override
    public UserResponse create(UserCreateRequest request) {
        var user = User.builder().name(request.name()).email(request.email()).build();

        return toResponse(userRepository.save(user));
    }

    @Override
    public UserResponse getById(UUID id) {
        return userRepository
                .findById(id)
                .map(this::toResponse)
                .orElseThrow(() -> new UserNotFoundException(id));
    }

    @Override
    public List<UserResponse> getAll() {
        return userRepository.findAll().stream().map(this::toResponse).toList();
    }

    @Override
    public UserResponse update(UUID id, UserUpdateRequest request) {
        var user = userRepository.findById(id).orElseThrow(() -> new UserNotFoundException(id));

        user.setName(request.name());
        user.setEmail(request.email());

        return toResponse(userRepository.save(user));
    }

    @Override
    public void delete(UUID id) {
        userRepository.deleteById(id);
    }

    private UserResponse toResponse(User user) {
        return UserResponse.builder()
                .id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .build();
    }
}

File: api/src/main/kotlin/dev/bitstachio/interdimensional_bazaar/user/service/UserServiceImpl.kt
package dev.bitstachio.interdimensional_bazaar.user.service

import dev.bitstachio.interdimensional_bazaar.user.domain.User
import dev.bitstachio.interdimensional_bazaar.user.dto.UserCreateRequest
import dev.bitstachio.interdimensional_bazaar.user.dto.UserResponse
import dev.bitstachio.interdimensional_bazaar.user.dto.UserUpdateRequest
import dev.bitstachio.interdimensional_bazaar.user.exception.UserDuplicateEmailException
import dev.bitstachio.interdimensional_bazaar.user.exception.UserNotFoundException
import dev.bitstachio.interdimensional_bazaar.user.repository.UserRepository
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.UUID

@Service
class UserServiceImpl(private val userRepository: UserRepository) : UserService {

	override fun create(request: UserCreateRequest): UserResponse {
		if (userRepository.existsByEmail(request.email)) {
			throw UserDuplicateEmailException(request.email)
		}
		val now = LocalDateTime.now()
		val user =
			User(
				firstName = request.firstName,
				lastName = request.lastName,
				email = request.email,
				passwordHash = request.passwordHash,
				role = request.role,
				phone = request.phone,
				createdAt = now,
			)
		return toResponse(userRepository.save(user))
	}

	override fun getById(id: UUID): UserResponse {
		return userRepository.findById(id).map(this::toResponse).orElseThrow { UserNotFoundException(id) }
	}

	override fun getAll(): List<UserResponse> {
		return userRepository.findAll().map(this::toResponse)
	}

	override fun update(id: UUID, request: UserUpdateRequest): UserResponse {
		val user = userRepository.findById(id).orElseThrow { UserNotFoundException(id) }
		if (userRepository.existsByEmailAndIdNot(request.email, id)) {
			throw UserDuplicateEmailException(request.email)
		}
		user.firstName = request.firstName
		user.lastName = request.lastName
		user.email = request.email
		user.passwordHash = request.passwordHash
		user.phone = request.phone
		return toResponse(userRepository.save(user))
	}

	override fun delete(id: UUID) {
		if (!userRepository.existsById(id)) {
			throw UserNotFoundException(id)
		}
		userRepository.deleteById(id)
	}

	private fun toResponse(user: User): UserResponse {
		return UserResponse(
			id = user.id!!,
			firstName = user.firstName,
			lastName = user.lastName,
			email = user.email,
			createdAt = user.createdAt,
			role = user.role,
			phone = user.phone,
		)
	}
}
